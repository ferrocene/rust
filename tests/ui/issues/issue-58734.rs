//@ revisions: edition2015 edition2021
//@[edition2015] edition: 2015
//@[edition2021] edition: 2021
trait Trait {
    fn exists(self) -> ();

    fn dyn_incompatible() -> Self;
}

impl Trait for () {
    fn exists(self) -> () {
    }

    fn dyn_incompatible() -> Self {
        ()
    }
}

fn main() {
    // dyn-compatible or not, this call is OK
    Trait::exists(());
    // no dyn-compatibility error
    Trait::nonexistent(());
    //[edition2015]~^ WARN trait objects without an explicit `dyn` are deprecated
    //[edition2015]~| WARN this is accepted in the current edition
    //[edition2015]~| ERROR the trait `Trait` is not dyn compatible
    //[edition2021]~^^^^ ERROR: expected a type, found a trait
}
