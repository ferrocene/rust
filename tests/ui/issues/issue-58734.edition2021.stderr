error[E0782]: expected a type, found a trait
  --> $DIR/issue-58734.rs:23:5
   |
LL |     Trait::nonexistent(());
   |     ^^^^^
   |
help: you can add the `dyn` keyword if you want a trait object
   |
LL |     <dyn Trait>::nonexistent(());
   |     ++++      +

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0782`.
